src/types.ts

```diff
@@ -0,0 +1,197 @@
+    options?: FlagEvaluationOptions
+  ): Promise<number>;
+
+  /**
+   * Get a number flag with additional details.
+   */
+  getNumberDetails(
+    flagKey: string,
+    defaultValue: number,
+    context?: EvaluationContext,
+    options?: FlagEvaluationOptions
+  ): Promise<EvaluationDetails<number>>;
+
+  /**
+   * Get an object (JSON) flag value.
+   */
+  getObjectValue<T extends object>(
+    flagKey: string,
+    defaultValue: T,
+    context?: EvaluationContext,
+    options?: FlagEvaluationOptions
+  ): Promise<T>;
+
+  /**
+   * Get an object (JSON) flag with additional details.
+   */
+  getObjectDetails<T extends object>(
+    flagKey: string,
+    defaultValue: T,
+    context?: EvaluationContext,
+    options?: FlagEvaluationOptions
+  ): Promise<EvaluationDetails<T>>;
+}
+
+/**
+ * Function which transforms the EvaluationContext to a type useful for the provider.
+ */
+export type ContextTransformer<T = unknown> = (context: EvaluationContext) => T;
+
+interface GenericProvider<T> {
+  name: string;
+
+  /**
+   * Resolve a boolean flag and it's evaluation details.
+   */
+  resolveBooleanEvaluation(
+    flagKey: string,
+    defaultValue: boolean,
+    transformedContext: T,
+    options: FlagEvaluationOptions | undefined
+  ): Promise<ResolutionDetails<boolean>>;
+
+  /**
+   * Resolve a string flag and it's evaluation details.
+   */
+  resolveStringEvaluation(
+    flagKey: string,
+    defaultValue: string,
+    transformedContext: T,
+    options: FlagEvaluationOptions | undefined
+  ): Promise<ResolutionDetails<string>>;
+
+  /**
+   * Resolve a numeric flag and it's evaluation details.
+   */
+  resolveNumberEvaluation(
+    flagKey: string,
+    defaultValue: number,
+    transformedContext: T,
+    options: FlagEvaluationOptions | undefined
+  ): Promise<ResolutionDetails<number>>;
+
+  /**
+   * Resolve and parse an object flag and it's evaluation details.
+   */
+  resolveObjectEvaluation<U extends object>(
+    flagKey: string,
+    defaultValue: U,
+    transformedContext: T,
+    options: FlagEvaluationOptions | undefined
+  ): Promise<ResolutionDetails<U>>;
+}
+
+export type NonTransformingProvider = GenericProvider<EvaluationContext>;
+
+export interface TransformingProvider<T> extends GenericProvider<T> {
+  contextTransformer: ContextTransformer<Promise<T> | T> | undefined;
+}
+
+/**
+ * Interface that providers must implement to resolve flag values for their particular
+ * backend or vendor.
+ *
+ * Implementation for resolving all the required flag types must be defined.
+ *
+ * Additionally, a ContextTransformer function that transforms the OpenFeature context to the requisite user/context/attribute representation (typeof T)
+ * may also be implemented. This function will run immediately before the flag value resolver functions, appropriately transforming the context.
+ */
+export type Provider<T extends EvaluationContext | unknown = EvaluationContext> = T extends EvaluationContext
```

Comment:
While I'm in favor of this approach for a context transformer, is it in the spec at this time? If not do we want to move it out until it is supported?

Reply:
https://github.com/open-feature/node-sdk/pull/8#issuecomment-1139637255

As Justin says, it's optional. I think with the recent conversations about the things like certain chars not being valid for particular vendor attribute keys, it might actually be more important to eventually make required.